### ISSUE TITLE
x/playground: use gVisor on linux/amd64 binaries instead of NaCl for sandboxing

### ISSUE THREAD

**Issue by andybons** [golang/go#25224]
Native Client deprecation has been announced in favor of Web Assembly. It's unclear what that means in terms of NaCl's future development and support.

I'd like us to investigate using gVisor as our sandbox mechanism for the playground. **This is not a statement that we should definitively switch over to gVisor.**

https://github.com/google/gvisor

@ysmolsky

**Comment by andybons** [comment]
This has come up again as attempting to upgrade the playground to 1.11 results in the following breakage during the NaCl time patching phase:

```
 ---> ed8ffc3e1836
Step 20/63 : RUN patch -p1 -d /usr/local/go </usr/local/playground/strict-time.patch
 ---> Running in 1f97414872b7
patching file src/runtime/os_nacl.go
Reversed (or previously applied) patch detected!  Assume -R? [n] 
Apply anyway? [n] 
Skipping patch.
1 out of 1 hunk ignored -- saving rejects to file src/runtime/os_nacl.go.rej
patching file src/runtime/sys_nacl_amd64p32.s
Reversed (or previously applied) patch detected!  Assume -R? [n] 
Apply anyway? [n] 
Skipping patch.
1 out of 1 hunk ignored -- saving rejects to file src/runtime/sys_nacl_amd64p32.s.rej
The command '/bin/sh -c patch -p1 -d /usr/local/go </usr/local/playground/strict-time.patch' returned a non-zero code: 1
```

**Comment by bradfitz** [comment]
That's @bcmills's patch. The old patch (enable-fake-time.patch) was designed to basically never bit rot but the new patch seems to have more context that makes it fragile. We should really upstream this and have it behind a bool or build tag so it's less likely to rot.

**Comment by bradfitz** [comment]
I've been playing around with this. I'm deciding between:

1. continuing to use App Engine Flex and launching gvisor's runsc "directly" (if one could consider the OCI interface direct). Unfortunately runsc still requires root today (for now: https://github.com/google/gvisor/issues/311) and App Engine Flex doesn't give you root at serving time (only for debugging). So doing this would require modifying runsc to not require root.

2. migrating off App Engine Flex and just using Docker with vanilla runsc. We'd need to change how we serve a bit. Probably an HTTP load balancer to a GCE VM instance group (perhaps running Container-Optimized OS, which already runs docker, and supports declarative yaml configs of which containers to run, including I believe privileged ones)

I think I'm leaning towards (2).

As part of this, we'd migrate to (2) well before the Go 1.14 release (which drops nacl), so the release process doesn't involve changing playground machinery during a release. We'd still support Go 1.13 nacl mode with the new infrastructure, and then the Go 1.14 release would just be changing a single line in its Dockerfile.

/cc @dmitshur @andybons @toothrot @prattmic

**Comment by toothrot** [comment]
@bradfitz Have you investigated using Cloud Run? The documentation seems to suggest that the runtime user is root: https://cloud.google.com/run/docs/tips#container-security

**Comment by bradfitz** [comment]
@toothrot, Cloud Run is already under gVisor itself. (and it doesn't let us tweak its sandbox config to be more restrictive, as we'd need). Cloud Run gives you root, but fake-ish constrained root. We definitely couldn't use gVisor's KVM mode ("platform") under it. We could probably just the ptrace mode (ptrace syscall is at least partially supported: https://gvisor.dev/docs/user_guide/compatibility/linux/amd64/). But once runsc went to do the other root-requiring stuff (setting up namespaces, etc), I suspect we'd trip over unimplemented system calls. e.g. I see on that system call list:

> Mount namespace (CLONE_NEWNS) not supported. Options CLONE_PARENT, CLONE_SYSVSEM not supported.

So basically, to use Cloud Run we'd need to do the same work hacking up runsc to run rootless that we'd have to do to run it on its current App Engine Flex environment.

That said, we should move many of our things to Cloud Run regardless, but I don't think it'd help us here.

@prattmic can correct me if indeed the ptrace mode of runsc would work unmodified under gvisor. That would be a happy surprise.

**Comment by bradfitz** [comment]
In case we do end up going the path of using some modified root-less runsc directly, without docker, this syzkaller code drives runsc directly:

https://github.com/google/syzkaller/blob/master/vm/gvisor/gvisor.go

**Comment by tv42** [comment]
For what it's worth, this doesn't seem to work right: gVisor runs see a lot of "Error communicating with remote server." and when it does work, stdout is not visible. https://play.golang.org/p/5_6CpdUjYY1

**Comment by bradfitz** [comment]
@tv42, it's not fully deployed yet.

**Comment by bradfitz** [comment]
Reopening until it's all deployed, though.

**Comment by bradfitz** [comment]
@toothrot, next up we need to run the frontend in a different network:

https://cloud.google.com/appengine/docs/flexible/go/using-shared-vpc

Have you done this before? (I haven't.)

Because right now it can't contact the internal load balancer.

We should test it first at a play-test.golang.org version and verify that it works (and doesn't break, say, memcache connections).

**Comment by bradfitz** [comment]
Oh, that page is for Shared VPC, not VPC. Maybe it's as simple as naming the network:

https://cloud.google.com/appengine/docs/flexible/go/reference/app-yaml#network_settings

Will try that in a test app.

**Comment by bradfitz** [comment]
Yup, it works with just the network declared. Looks like I'm missing a docker pull on start-up, though, causing stderr spam:

<img width="642" alt="Screen Shot 2020-01-09 at 11 47 21 AM" src="https://user-images.githubusercontent.com/2621/72099695-fb137280-32d5-11ea-975e-0138ea506913.png">


And looks like memcache is broken (as expected, as the memcache VM is running in a different VPC now). But that's easy to fix.

**Comment by toothrot** [comment]
Nice, I'm glad that worked out. For the record, I haven't tried using Flex on a shared VPC before, but I'm glad it works fine with its own VPC.

**Comment by bradfitz** [comment]
Created a new memcached VM (named `memcached-play-golang` instead old `memcached`) in the right VPC and verified it's working:

```
$ (echo stats; echo quit) | nc memcached-play-golang 11211 2>&1 | grep -v " 0"
STAT pid 25186
STAT uptime 1179
STAT time 1578602058
STAT version 1.4.33
STAT libevent 2.0.21-stable
STAT pointer_size 64
STAT curr_connections 5
STAT total_connections 19
STAT connection_structures 6
STAT reserved_fds 20
STAT cmd_get 5
STAT cmd_set 1
STAT get_hits 4
STAT get_misses 1
STAT bytes_read 1692
STAT bytes_written 20337
STAT limit_maxbytes 3670016000
STAT accepting_conns 1
STAT threads 4
STAT hash_power_level 16
STAT hash_bytes 524288
STAT bytes 1164
STAT curr_items 1
STAT total_items 1
END
```

Will send a CL to change the `app.yaml` config.

**Comment by bradfitz** [comment]
Something's also up with the health checks:

![Uploading Screen Shot 2020-01-09 at 12.56.00 PM.png…]()

I might've defined them incorrectly in the `sandbox.tf` terraform file.

**Comment by toothrot** [comment]
@bradfitz As far as I am aware, this is in a state where it is no longer a release-blocker for 1.14, but not yet ready to be closed, is that correct?

**Comment by bradfitz** [comment]
Probably, yeah. It hasn't been load tested or monitored or productionized (health check configs) quite enough yet probably. I'll leave for you to decide.

**Comment by odeke-em** [comment]
Thank you Brad and Alex! @toothrot, we’ve removed the “release-blocker” label but this issue is still in the Go1.14 milestone. Should we perhaps add it to Go.1.15 or Backlog, and repurpose it for load testing, monitoring and production hardening? Or close this one and open a separate issue for the mentioned goals?

**Comment by smasher164** [comment]
I noticed although `runtime.Version` reports `go1.14.2` (https://play.golang.org/p/Ztyu2FJaajl), the About page says 
```
The playground uses the latest stable release of Go.
The current version is go1.13.9.
```

**Comment by toothrot** [comment]
@smasher164 Thanks! That will be fixed when this issue is resolved.

**Comment by toothrot** [comment]
@smasher164 The issue you mention should be resolved.

**Comment by toothrot** [comment]
I should mention here that one of the biggest issues in resolving this is lack of TryBot support. I will see if there is any easy way to make that happen.

**Comment by dmitshur** [comment]
@toothrot Issue #24823 is relevant.

**Comment by toothrot** [comment]
Cross-linking this comment mentioning our new sandbox metrics for posterity: https://github.com/golang/go/issues/38530#issuecomment-620228738

**Comment by toothrot** [comment]
I misunderstood in the referenced issue #38530: you can use custom metrics for autoscaling: https://cloud.google.com/compute/docs/autoscaler/scaling-stackdriver-monitoring-metrics#custom_metrics

QPS or container_count could be good candidates. We'd have to update our target to be a GCE instance rather than a generic_task for the metric we choose.

**Comment by toothrot** [comment]
I think this is effectively finished.
